### PR TITLE
Create user model and migration

### PR DESCRIPTION
**Pull request by kau-wtag**
Here I added User model and migration.

Title: Setting Up User Model, Database Structure, and Tests

Background:
I’m starting to build an event management system using Ruby on Rails. To make this happen, we need to create a way for users to interact with the system. This involves creating a User model, which will store information about the users, like their name and email. Additionally, it's important to ensure that our User model works correctly by writing comprehensive tests.

Goal:

- Create User Model: Make a blueprint for how user data will be stored.
- Create migration: Create a structure in the database to hold user information.
- Add Tests for User Model: Write RSpec tests to validate the functionality and integrity of the User model.

Acceptance:

- User Model Created: We'll have a User model ready to handle user data.
- Database Structure Defined: The database will be set up to store user information securely and efficiently.
- Tests Added and Passed: RSpec tests will be in place to ensure the User model functions as expected, and these tests will pass.

Story: [https://trello.com/c/3nDgtT0B](https://trello.com/c/3nDgtT0B)

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  validates :name, presence: true
+  validates :email, format: { with: /\S+@\S+/ }, uniqueness: { case_sensitive: false }
```
Comment (sku-wtag) [inline app/models/user.rb]: Rails has a default REGEX for matching emails. You can use this.

```suggestion
  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: { case_sensitive: false }
```

**Review by asis-wtag — APPROVED** [review]

File: db/migrate/20240609101057_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[7.1]
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :email
```
Comment (sku-wtag) [inline db/migrate/20240609101057_create_users.rb]: Can name and email be null?

File: db/migrate/20240609101057_create_users.rb
Comment (sal-wtag) [inline db/migrate/20240609101057_create_users.rb:1]: @udk-wtag bhai, according to the migration file, `name` and `email` can be `null` at the db level, right?

File: db/migrate/20240609101057_create_users.rb (hunk 2 above)
Comment (sal-wtag) [inline db/migrate/20240609101057_create_users.rb]: @udk-wtag bhai, according to the migration file, are we allowing `null` values for `name` and `email` at the db level?

File: test/models/user_test.rb
Comment (sal-wtag) [inline test/models/user_test.rb:1]: @udk-wtag bhai, we prefer `RSpec` for testing, not the default `minitest` framework.

**Review by sal-wtag — COMMENTED** [review]
@udk-wtag Nice work!

## Few Suggestions
- Try to use both model and db level constraints to ensure data integrity.
- Prefer `RSpec` over default testing framework.

File: app/models/user.rb (hunk 1 above)
Comment (kau-wtag) [reply app/models/user.rb]: @sku-wtag bhai, I have updated it.

File: db/migrate/20240609101057_create_users.rb (hunk 2 above)
Comment (kau-wtag) [reply db/migrate/20240609101057_create_users.rb]: @sal-wtag bhai, I have only added validation at model level. I am working on to add validation also at db lebel.

File: test/models/user_test.rb
Comment (kau-wtag) [reply test/models/user_test.rb:1]: @sal-wtag bhai, Added RSpec.

**Review by sal-wtag — APPROVED** [review]

File: db/migrate/20240609101057_create_users.rb (hunk 2 above)
Comment (kau-wtag) [reply db/migrate/20240609101057_create_users.rb]: Done it bhai

**Review by sku-wtag — APPROVED** [review]

File: .rspec [hunk 3]
```diff
@@ -0,0 +1 @@
+--require spec_helper
```
Comment (SKH-WTAG) [inline .rspec]: Import rails_helper instead of spec_helper

File: db/migrate/20240612102426_add_constraints_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class AddConstraintsToUsers < ActiveRecord::Migration[7.1]
+  def change
+    change_column_null :users, :name, false
+    change_column_null :users, :email, false
+    change_column_null :users, :password_digest, false
```
Comment (SKH-WTAG) [inline db/migrate/20240612102426_add_constraints_to_users.rb]: Why do we need seperate migration file to change column?

File: db/migrate/20240609101057_create_users.rb (hunk 2 above)
Comment (SKH-WTAG) [inline db/migrate/20240609101057_create_users.rb]: Use `citext`

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  validates :name, presence: true
+  validates :email, presence: true, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: { case_sensitive: false }
+  validates :password, length: { minimum: 6 }
```
Comment (SKH-WTAG) [inline app/models/user.rb]: Will it validate presence of password?

File: spec/factories/users.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :user do
+    name { Faker::Name.name }
```
Comment (SKH-WTAG) [inline spec/factories/users.rb]: use FFaker

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  before do
+    User.destroy_all
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Why do we need to destroy all? rspec should cleanup automatically.

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  context 'validations' do
+    it 'is valid with valid attributes' do
+      user = build(:user)
+      expect(user).to be_valid
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Use `shoulda-matchers`

File: test/fixtures/users.yml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (SKH-WTAG) [inline test/fixtures/users.yml]: Do we need this?

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
```
Comment (SKH-WTAG) [inline test/models/user_test.rb]: ditto

File: .rspec (hunk 3 above)
Comment (kau-wtag) [reply .rspec]: Bhai, done.

File: db/migrate/20240612102426_add_constraints_to_users.rb (hunk 4 above)
Comment (kau-wtag) [reply db/migrate/20240612102426_add_constraints_to_users.rb]: Bhai, i forgot to add necessary constraints on 1st migration, So I created 2nd migration to add the constraints.

File: db/migrate/20240609101057_create_users.rb (hunk 2 above)
Comment (kau-wtag) [reply db/migrate/20240609101057_create_users.rb]: Bhai, used citext for email.

File: spec/factories/users.rb (hunk 6 above)
Comment (kau-wtag) [reply spec/factories/users.rb]: Bhai, Done.

File: test/fixtures/users.yml (hunk 9 above)
Comment (kau-wtag) [reply test/fixtures/users.yml]: No bhai, Removed it.

File: spec/models/user_spec.rb (hunk 8 above)
Comment (kau-wtag) [reply spec/models/user_spec.rb]: Bhai, done.

File: test/models/user_test.rb (hunk 10 above)
Comment (kau-wtag) [reply test/models/user_test.rb]: Bhai, deleted the test folder.

File: spec/models/user_spec.rb (hunk 7 above)
Comment (kau-wtag) [reply spec/models/user_spec.rb]: Bhai, done.

File: app/models/user.rb (hunk 5 above)
Comment (kau-wtag) [reply app/models/user.rb]: No bhai, To validate presence of password. presence: true should be add.

File: Gemfile [hunk 11]
```diff
@@ -11,10 +11,14 @@ gem 'sprockets-rails'
...
 gem 'turbo-rails'
 gem 'tzinfo-data', platforms: %i[windows jruby]
+gem 'bcrypt', '~> 3.1.7'
 
 group :development, :test do
   gem 'dotenv-rails'
   gem 'rspec-rails'
+  gem 'factory_bot_rails', '~> 6.2'
+  gem 'ffaker', '~> 2.21'
+  gem 'shoulda-matchers', '~> 6.2'
```
Comment (SKH-WTAG) [inline Gemfile]: Do we need shoulda-machers in `development`?

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
+require 'rails_helper'
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Do we need this?

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+RSpec.describe User, type: :model do
+  describe 'validations' do
+    subject { build(:user) }
+
+    it { should validate_presence_of(:name) }
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: we use `is_expected.to`

File: spec/models/user_spec.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      user = build(:user, password: 'password123', password_confirmation: 'different')
+      expect(user).not_to be_valid
+      expect(user.errors[:password_confirmation]).to include("doesn't match Password")
+    end
+
+    it 'is not valid with a short password' do
+      user = build(:user, password: 'short', password_confirmation: 'short')
+      expect(user).not_to be_valid
+      expect(user.errors[:password]).to include("is too short (minimum is 6 characters)")
+    end
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Use shoulda-machers provided methods to check

File: Gemfile (hunk 11 above)
Comment (kau-wtag) [reply Gemfile]: No Bhai, we can move it into test block.

File: spec/models/user_spec.rb (hunk 13 above)
Comment (kau-wtag) [reply spec/models/user_spec.rb]: Done, bhai.

File: spec/models/user_spec.rb (hunk 14 above)
Comment (kau-wtag) [reply spec/models/user_spec.rb]: Bhai, Done.

File: spec/models/user_spec.rb (hunk 12 above)
Comment (kau-wtag) [reply spec/models/user_spec.rb]: Removed, bhai.

**Review by SKH-WTAG — APPROVED** [review]

**Pull request merged**